### PR TITLE
fix: correctly cancel failed builds

### PR DESCRIPTION
**Pull request by rgrinberg**
Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

File: src/dune_engine/process.ml [hunk 1]
```diff
@@ -843,32 +843,30 @@ let run_internal ?dir ~(display : Display.t) ?(stdout_to = Io.stdout)
...
+      (* if the cancellation token was fired, then we:
+
+         1) aren't interested in printing the output from the cancelled job
+
+         2) allowing callers to continue work with the already stale value
+         we're about to return. We reuse [Already_reported] to signal that
+         this exception is propagated without being reported. It's not the
+         original intention of [Already_reported] but it works adequately
+         here. *)
+      let+ () = Scheduler.abort_if_build_was_cancelled in
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_engine/process.ml]: I cleaned up the comment and also just raise the cancellation exception directly. Going through the scheduler is pointless here and we can remove this helper.

**Comment by dkalinichenko-js** [comment]
Just tested internally, the fix works.

**Comment by rgrinberg** [comment]
@emillon this is a good candidate for 3.8.3 as well

**Comment by rgrinberg** [comment]
I don't know how you feel about it, but perhaps 3.8.4?

**Comment by emillon** [comment]
Definitely 3.8.4 if there's one, but I hope 3.9.0 (tomorrow) will be a painless upgrade so this will be a good way to deliver this fix. If it's difficult to release or has regressions that would warrant a 3.8.4.

**Review by snowleopard — APPROVED** [review]
Thanks! This fixes the problem on our end.

**Comment by snowleopard** [comment]
Just to say, I think this bug fix is pretty important: the watch mode was basically unusable without it for us internally.

**Comment by rgrinberg** [comment]
Turns out that the buggy commit never made it into any released versions. So there's no need for any bug fix releases.

**Comment by emillon** [comment]
that's a relief. thanks

**Pull request merged**